lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -262,6 +268,21 @@ defmodule Mix.Compilers.Elixir do
     Mix.shell.info "Compiling #{source} (it's taking more than #{threshold}s)"
   end
 
+  defp each_warning(pid, cwd, file, line, warning) do
```

Comment:
Hrm, I think a better approach would be to change the agent state to be a tuple with three elements: `{modules, sources, warnings}` where `warnings` is a map with the source as key and a list of warnings as data. This way, all you need to do in the code above is to `Map.pop(warnings, source_path, [])`. It will also be much faster for codebase with multiple warnings.